File: lib/data/models/library_response.dart
class LibraryResponse {
  const LibraryResponse({
    required this.systemId,
    required this.systemName,
    required this.libKey,
    required this.libId,
    required this.shortName,
    required this.formalName,
    required this.address,
    required this.pref,
    required this.city,
    required this.category,
    this.urlPc,
    this.post,
    this.tel,
    this.geocode,
  });

  final String systemId;
  final String systemName;
  final String libKey;
  final String libId;
  final String shortName;
  final String formalName;
  final String? urlPc;
  final String address;
  final String pref;
  final String city;
  final String? post;
  final String? tel;
  final String? geocode;
  final String category;

  factory LibraryResponse.fromJson(Map<String, dynamic> json) {
    return LibraryResponse(
      systemId: json['systemid'] as String? ?? '',
      systemName: json['systemname'] as String? ?? '',
      libKey: json['libkey'] as String? ?? '',
      libId: json['libid'] as String? ?? '',
      shortName: json['short'] as String? ?? '',
      formalName: json['formal'] as String? ?? '',
      urlPc: json['url_pc'] as String?,
      address: json['address'] as String? ?? '',
      pref: json['pref'] as String? ?? '',
      city: json['city'] as String? ?? '',
      post: json['post'] as String?,
      tel: json['tel'] as String?,
      geocode: json['geocode'] as String?,
      category: json['category'] as String? ?? '',
    );
  }
}

File: src/data/models/libraryResponse.ts
/**
 * Model for the Calil `/library` API response entry.
 * Mirrors the Dart `LibraryResponse`.
 *
 * The Calil API uses lowercase snake-case keys:
 * `systemid, systemname, libkey, libid, short, formal, url_pc,
 *  address, pref, city, post, tel, geocode, category`.
 */
export interface LibraryResponse {
  readonly systemId: string;
  readonly systemName: string;
  readonly libKey: string;
  readonly libId: string;
  readonly shortName: string;
  readonly formalName: string;
  readonly urlPc?: string;
  readonly address: string;
  readonly pref: string;
  readonly city: string;
  readonly post?: string;
  readonly tel?: string;
  readonly geocode?: string;
  readonly category: string;
}

function stringOrDefault(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

function optionalString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

export function libraryResponseFromJson(
  json: Record<string, unknown>,
): LibraryResponse {
  return {
    systemId: stringOrDefault(json.systemid),
    systemName: stringOrDefault(json.systemname),
    libKey: stringOrDefault(json.libkey),
    libId: stringOrDefault(json.libid),
    shortName: stringOrDefault(json.short),
    formalName: stringOrDefault(json.formal),
    urlPc: optionalString(json.url_pc),
    address: stringOrDefault(json.address),
    pref: stringOrDefault(json.pref),
    city: stringOrDefault(json.city),
    post: optionalString(json.post),
    tel: optionalString(json.tel),
    geocode: optionalString(json.geocode),
    category: stringOrDefault(json.category),
  };
}
